pygmt/datatypes/grid.py

```diff
@@ -169,16 +173,14 @@ def to_dataarray(self) -> xr.DataArray:
         # Get dimensions and their attributes from the header.
         dims, dim_attrs = header.dims, header.dim_attrs
         # The coordinates, given as a tuple of the form (dims, data, attrs)
-        coords = [
-            (dims[0], self.y[: header.n_rows], dim_attrs[0]),
-            (dims[1], self.x[: header.n_columns], dim_attrs[1]),
-        ]
+        x = np.ctypeslib.as_array(self.x, shape=(header.n_columns,)).copy()
+        y = np.ctypeslib.as_array(self.y, shape=(header.n_rows,)).copy()
+        coords = [(dims[0], y, dim_attrs[0]), (dims[1], x, dim_attrs[1])]
 
         # The data array without paddings
+        data = np.ctypeslib.as_array(self.data, shape=(header.my, header.mx)).copy()
```

Comment:
> Also need to be cautious that we need to make a copy of the ctypes/numpy array, since the memory of the array is allocated by GMT and will be freed when exiting the session.

Trying to understand this sentence a bit more. So we can't assign ownership of the array data from GMT to numpy/xarray, and need to make a copy to ensure that exiting the GMT session doesn't crash the program? It doesn't look like we needed to explicitly make a copy before, so why is it needed now?

Reply:
> So we can't assign ownership of the array data from GMT to numpy/xarray, and need to make a copy to ensure that exiting the GMT session doesn't crash the program? 

Yes, the arrays allocated by GMT will be freed when exiting the GMT session. This can be verified by the example below. In the GMT session, we can correctly access the `grid.x` array, but outside the session, `grid.x` reports a "NULL pointer access" error. 
```python
In [6]: from pygmt.clib import Session

In [7]: with Session() as lib:
   ...:     with lib.virtualfile_out(kind="grid") as voutgrd:
   ...:         lib.call_module("read", ["@static_earth_relief.nc", voutgrd, "-Tg"])
   ...:         grid = lib.read_virtualfile(voutgrd, kind="grid").contents
   ...:         header = grid.header.contents
   ...:         print(grid.x[:header.n_columns])
   ...: print(grid.x[0:5])
[-54.5, -53.5, -52.5, -51.5, -50.5, -49.5, -48.5, -47.5]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[7], line 7
      5         header = grid.header.contents
      6         print(grid.x[:header.n_columns])
----> 7 print(grid.x[0:5])

ValueError: NULL pointer access
```

> It doesn't look like we needed to explicitly make a copy before, so why is it needed now?
As explained in the top post, in the previous version, we first access the ctypes array via the list slice syntax (i.e., `ctypes_array[:120]`) and then convert the list slice to a numpy array by `np.reshape`. The list slice doesn't make a new copy of the array memory, but the list->ndarray conversion changes the data dtype from `float32` (the default dtype for GMT grid) to `float64` (the default dtype for numpy). I think in this conversion, numpy makes a new copy of the array:
```
array1 = np.reshape(ctypes_array[:120], (10, 12))
```

Reply:
> > It doesn't look like we needed to explicitly make a copy before, so why is it needed now?
> 
> As explained in the top post, in the previous version, we first access the ctypes array via the list slice syntax (i.e., `ctypes_array[:120]`) and then convert the list slice to a numpy array by `np.reshape`. The list slice doesn't make a new copy of the array memory, but the list->ndarray conversion changes the data dtype from `float32` (the default dtype for GMT grid) to `float64` (the default dtype for numpy). I think in this conversion, numpy makes a new copy of the array:
> 
> ```
> array1 = np.reshape(ctypes_array[:120], (10, 12))
> ```

Here is a test script:
```
In [22]: import ctypes

In [23]: import numpy as np

In [24]: ctypes_array = (ctypes.c_uint8 * 12)(*range(12))

In [25]: array1 = np.reshape(ctypes_array[:12], (3, 4))

In [26]: ctypes_array[:12]
Out[26]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [27]: array1
Out[27]:
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [28]: ctypes_array[:12] = list(range(100, 112))

In [29]: ctypes_array
Out[29]: <__main__.c_ubyte_Array_12 at 0x7f88d81b1fd0>

In [30]: ctypes_array[:12]
Out[30]: [100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111]

In [31]: array1
Out[31]:
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])
```
```

Reply:
> but the list->ndarray conversion changes the data dtype from `float32` (the default dtype for GMT grid) to `float64` (the default dtype for numpy). I think in this conversion, numpy makes a new copy of the array:

Ok, this part makes sense, so we were implicitly doing a copy of the array before, whereas now it's more explicit via a `.copy()` call.

> the arrays allocated by GMT will be freed when exiting the GMT session.

Thanks for showing the "NULL pointer access" exception. We probably don't want the pointer to remain active after the session is closed since that would be a memory leak, but is there a way for GMT to pass ownership of the array data to NumPy before exiting the session, so that the memory is then managed by Python (and its garbage collector)?

Reply:
> is there a way for GMT to pass ownership of the array data to NumPy before exiting the session, so that the memory is then managed by Python (and its garbage collector)?

I thought the API function `GMT_Set_AllocMode` may help, but I just tried it locally and it doesn't work as I expected. Maybe another upstream bug or something else that needs to debug the C codes.

Reply:
From this post https://stackoverflow.com/questions/28760273/python-ctypes-and-garbage-collection and some answers from ChatGPT, it seems even if it's possible to tell GMT to not free the memory, the Python side is still responsible to free the memory explicitly. Otherwise, there may be memory leaks.

Reply:
Ok, so we either need to:

1. Allocate the memory space in Python and have GMT C use it with `GMT_Set_AllocMode` so that memory is still managed by Python
2. Allocate memory space in GMT C, pass ownership to Python, and ensure that Python handles free-ing the memory when no longer needed

Maybe something to think about longer term if we want to avoid copies :slightly_smiling_face: